### PR TITLE
[CIS-1081] Expose models on Events

### PR DESCRIPTION
**Pull request by b-onc**
Missing:
- Changelog
- Tests

The way this works, we put a helper object to the event, and we put closures to helper object to fetch the data from DB. The closure is ran lazily when users access the fields (or never ran)

### CODE REVIEW

File: Sources/StreamChat/WebSocketClient/Events/Event.swift [hunk 1]
```diff
@@ -2,16 +2,94 @@
...
+        (self as? EventWithPayload)?.savedData?.unreadCount
+    }
+    
+    public func message() -> ChatMessage? {
+        if let eventWithPayload = self as? EventWithPayload {
+            if let savedData = eventWithPayload.savedData {
+                if let message = savedData.message {
+                    return message
+                } else {
+                    print("### no message")
```
Comment (b-onc) [inline Sources/StreamChat/WebSocketClient/Events/Event.swift:45]: I forgot these during debugging 😓

File: Sources/StreamChat/Workers/EventNotificationCenter.swift [hunk 2]
```diff
@@ -34,6 +34,46 @@ class EventNotificationCenter: NotificationCenter {
...
+            )
+            event = channelSpecificEvent
+        case var currentUserEvent as CurrentUserEvent:
+            currentUserEvent.savedData = SavedEventData(
+                currentUser: { [weak database] in database?.viewContext.currentUser?.asModel() },
+                underlyingContext: database.viewContext
+            )
+            event = currentUserEvent
+        default:
+            break
```
Comment (tbarbugli) [inline Sources/StreamChat/Workers/EventNotificationCenter.swift:74]: we need to handle custom events as well

File: Sources/StreamChatUI/ChatMessageList/ChatChannelHeaderView.swift [hunk 3]
```diff
@@ -127,7 +127,7 @@ open class ChatChannelHeaderView:
 
     open func channelController(
         _ channelController: ChatChannelController,
-        didReceiveMemberEvent: MemberEvent
+        didReceiveMemberEvent: Event
```
Comment (tbarbugli) [inline Sources/StreamChatUI/ChatMessageList/ChatChannelHeaderView.swift:130]: why this change?

File: DemoApp/ChatPresenter.swift [hunk 4]
```diff
@@ -176,22 +176,48 @@ class DemoChatChannelListRouter: ChatChannelListRouter {
...
     }
 
     @objc open func didTapCreateNewChannel(_ sender: Any) {
         (router as! DemoChatChannelListRouter).showCreateNewChannelFlow()
     }
+    
+    func eventsController(_ controller: EventsController, didReceiveEvent event: Event) {
+        switch event {
+        case let event as MessageNewEvent:
+            if let message = event.message(), let channel = event.channel() {
```
Comment (dmigach) [inline DemoApp/ChatPresenter.swift:206]: The whole approach totally makes sense to me, the only thing I'd maybe change is turn `message()`, `channel()` etc on `Event` into computed properties so when you use it the syntax is more natural, unless you think there is some value in explicitly showing that there is computation under the hood

File: Sources/StreamChatUI/ChatMessageList/ChatChannelHeaderView.swift (hunk 3 above)
Comment (b-onc) [reply Sources/StreamChatUI/ChatMessageList/ChatChannelHeaderView.swift:130]: `MemberEvent` needs to have `savedData`, which is defined in `EventWithPayload`. `EventWithPayload` is `internal`, so we either expose `EventWithPayload` (bad idea) or we change these params `MemberEvent` -> `Event` so we don't need to expose payloads. The good part is, this change is basically free for the user, since the user needs to cast `MemberEvent` to a specific event (`MemberAddedEvent` or `MemberRemovedEvent`) anyway

File: DemoApp/ChatPresenter.swift (hunk 4 above)
Comment (b-onc) [reply DemoApp/ChatPresenter.swift:206]: if those are computed properties instead of functions, we have to implement them for all conformers of `Event` type. when they're functions, we can have a default implementation and all conformers will get the default implementation
